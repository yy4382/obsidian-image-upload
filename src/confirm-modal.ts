export type PromiseHandler = ReturnType<
	typeof createPromiseWithResolver
>["handler"];
import { Modal, Setting, type App } from "obsidian";

export function createPromiseWithResolver() {
	let resolvePromise: (value: unknown) => void;
	let rejectPromise: (reason: unknown) => void;

	const promise = new Promise((resolve, reject) => {
		resolvePromise = resolve;
		rejectPromise = reject;
	});

	const resolver = (value: unknown) => {
		resolvePromise(value);
	};

	const rejecter = (reason: unknown) => {
		rejectPromise(reason);
	};

	return { promise, handler: { resolver, rejecter } };
}

export class ConfirmModal extends Modal {
	private handler: PromiseHandler;
	private needReject: boolean;
	constructor(app: App, handler: PromiseHandler) {
		super(app);
		this.handler = handler;
		this.needReject = true;
	}

	confirm(value?: unknown): void {
		this.handler.resolver(value);
		this.close();
	}
	cancel(reason?: Error): void {
		if (this.needReject) {
			this.handler.rejecter(reason ?? new Error("User aborted"));
		}
		this.needReject = false;
	}
}

export class DeleteConfirmModal extends ConfirmModal {
	files: string[];
	constructor(app: App, handler: PromiseHandler, file: string[]) {
		super(app, handler);
		this.files = file;
	}
	onOpen(): void {
		const { contentEl } = this;
		contentEl.createEl("h2", {
			text: "Are you sure you want to delete these files?",
		});
		const ul = contentEl.createEl("ul");
		for (const file of this.files) {
			ul.createEl("li", { text: file });
		}

		new Setting(contentEl).addButton((button) => {
			button.setButtonText("Confirm").onClick(() => {
				this.confirm();
			});
		});
	}
	onClose(): void {
		this.cancel();
	}
}
