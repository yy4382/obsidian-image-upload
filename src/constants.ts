import type { Settings } from "./settings";

export const DEFAULT_SETTINGS: Settings = {
	s3: {
		endpoint: "",
		accKeyId: "",
		secretAccKey: "",
		bucket: "",
		keyTemplate:
			"obsidian/{{year}}/{{month}}/{{day}}/{{base62_of_ms_from_day_start}}-{{random2}}.{{extension}}",
		region: "",
		forcePathStyle: false,
		publicUrl: "",
	},
	customUploaderClass: "",
	uploadExt: "png jpg jpeg gif webp",
	useSystemTrash: false,
};
