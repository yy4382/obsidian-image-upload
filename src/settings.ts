import { type App, PluginSettingTab, Setting } from "obsidian";
import type ImageUploadPlugin from "./main";

export type Settings = {
	s3: {
		endpoint: string;
		accKeyId: string;
		secretAccKey: string;
		bucket: string;
		keyTemplate: string;
		region: string;
		forcePathStyle: boolean;
		publicUrl: string;
	};
	uploadExt: string;
	useSystemTrash: boolean;
};

export { DEFAULT_SETTINGS } from "./constants";

export class SettingsTab extends PluginSettingTab {
	plugin: ImageUploadPlugin;

	constructor(app: App, plugin: ImageUploadPlugin) {
		super(app, plugin);
		this.plugin = plugin;
	}

	display(): void {
		const { containerEl } = this;

		containerEl.empty();

		new Setting(containerEl).setHeading().setName("S3 Settings");
		this.addTextSetting(
			"Endpoint",
			"Endpoint of the S3 service",
			() => this.plugin.settings.s3.endpoint,
			(value) => {
				this.plugin.settings.s3.endpoint = value;
			},
		);
		this.addTextSetting(
			"Access Key ID",
			"Access Key ID for the S3 service",
			() => this.plugin.settings.s3.accKeyId,
			(value) => {
				this.plugin.settings.s3.accKeyId = value;
			},
		);
		this.addTextSetting(
			"Secret Access Key",
			"Secret Access Key for the S3 service",
			() => this.plugin.settings.s3.secretAccKey,
			(value) => {
				this.plugin.settings.s3.secretAccKey = value;
			},
		);
		this.addTextSetting(
			"Bucket",
			"Bucket to upload to",
			() => this.plugin.settings.s3.bucket,
			(value) => {
				this.plugin.settings.s3.bucket = value;
			},
		);
		this.addTextSetting(
			"Key Template",
			"Template for the key to use in S3",
			() => this.plugin.settings.s3.keyTemplate,
			(value) => {
				this.plugin.settings.s3.keyTemplate = value;
			},
		);
		this.addTextSetting(
			"Region",
			"Region of the S3 service",
			() => this.plugin.settings.s3.region,
			(value) => {
				this.plugin.settings.s3.region = value;
			},
		);
		this.addTextSetting(
			"Public URL",
			"Public URL of the S3 service",
			() => this.plugin.settings.s3.publicUrl ?? "",
			(value) => {
				this.plugin.settings.s3.publicUrl = value;
			},
		);
		new Setting(containerEl)
			.setName("Force Path Style")
			.setDesc("Whether to use path-style addressing")
			.addToggle((toggle) => {
				toggle
					.setValue(this.plugin.settings.s3.forcePathStyle)
					.onChange(async (value) => {
						this.plugin.settings.s3.forcePathStyle = value;
						await this.plugin.saveSettings();
					});
			});
		new Setting(containerEl).setHeading().setName("General Settings");
		new Setting(containerEl)
			.setName("Upload Extensions")
			.setDesc("Extensions to upload")
			.addText((text) =>
				text
					.setValue(this.plugin.settings.uploadExt)
					.onChange(async (value) => {
						this.plugin.settings.uploadExt = value;
						await this.plugin.saveSettings();
					}),
			);
		new Setting(containerEl).setName("Use System Trash").addToggle((toggle) => {
			toggle
				.setValue(this.plugin.settings.useSystemTrash)
				.onChange(async (value) => {
					this.plugin.settings.useSystemTrash = value;
					await this.plugin.saveSettings();
				});
		});
	}
	addTextSetting(
		name: string,
		desc: string,
		get: () => string,
		set: (value: string) => void,
	): void {
		new Setting(this.containerEl)
			.setName(name)
			.setDesc(desc)
			.addText((text) =>
				text.setValue(get()).onChange(async (value) => {
					set(value);
					await this.plugin.saveSettings();
				}),
			);
	}
}
