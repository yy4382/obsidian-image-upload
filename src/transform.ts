import type {
	EmbedCache,
	LinkCache,
	Pos,
	App,
	ReferenceCache,
	Notice,
	Vault,
	MetadataCache,
} from "obsidian";
import type { Settings } from "./settings";
import type { PTFile } from "./main";

export type TransformCtx = {
	settings: Settings;
	uploader: (binary: ArrayBuffer, tFile: PTFile) => Promise<string>;
	resolveLink: MetadataCache["getFirstLinkpathDest"];
	readBinary: Vault["readBinary"];
	notice: (...args: ConstructorParameters<typeof Notice>) => void;
};

export async function transform(
	ctx: TransformCtx,
	content: string,
	meta: {
		selfPath: string;
		links: LinkCache[];
		embeds: EmbedCache[];
	},
): Promise<[string, string[]]> {
	const process = async (link: ReferenceCache, isEmbed: boolean) => {
		const targetFile = ctx.resolveLink(link.link.split("#")[0], meta.selfPath);
		if (!targetFile) {
			ctx.notice(`File "${link.original}" not found in vault.`);
			return;
		}
		if (!ctx.settings.uploadExt.split(" ").includes(targetFile.extension)) {
			ctx.notice(
				`File "${targetFile.basename}" won't be uploaded for having extension "${targetFile.extension}".
Go to settings to configure that.`,
			);
			return;
		}
		const content = await ctx.readBinary(targetFile);
		let url: string;
		try {
			url = await ctx.uploader(content, targetFile);
		} catch (e) {
			console.error(e);
			ctx.notice(`Failed to upload file: ${targetFile.path}`);
			return;
		}
		const newLink = `${isEmbed ? "!" : ""}[${link.displayText ?? link.link}](${url})`;
		return [link.position, newLink, targetFile.path] as [Pos, string, string];
	};

	const actions = (
		await Promise.all([
			...meta.links.map((link) => process(link, false)),
			...meta.embeds.map((embed) => process(embed, true)),
		])
	).filter((x) => x !== undefined);

	const newContent = replace(content, actions);
	const uploadedPaths = dedupe(actions.map(([, , path]) => path));
	return [newContent, uploadedPaths];
}

function dedupe<T>(arr: T[]): T[] {
	return [...new Set(arr)];
}

function replace(
	content: string,
	replaceList: [Pos, string, unknown][],
): string {
	replaceList.sort((a, b) => b[0].start.offset - a[0].start.offset);
	return replaceList.reduce((acc, [pos, str]) => {
		return acc.slice(0, pos.start.offset) + str + acc.slice(pos.end.offset);
	}, content);
}

export function getDeletingFiles(
	uploadedPaths: string[],
	resolvedLinks: App["metadataCache"]["resolvedLinks"],
	sourcePath: string,
): string[] {
	const deletingFiles = new Set<string>();
	for (const path of uploadedPaths) {
		let found = false;
		for (const source in resolvedLinks) {
			if (source === sourcePath) continue;
			if (Object.keys(resolvedLinks[source]).includes(path)) {
				found = true;
				break;
			}
		}
		if (!found) deletingFiles.add(path);
	}
	return [...deletingFiles];
}
