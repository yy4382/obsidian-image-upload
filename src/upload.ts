import { S3Client, PutObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import type { requestUrl, TFile } from "obsidian";
import type { Settings } from "./settings";
import mime from "mime";
import { encode as encode62 } from "base62";
import type { PTFile } from "./main";

export type UploadCtx = {
	settings: Settings;
	requestUrl: typeof requestUrl;
};

export async function upload(
	binary: ArrayBuffer,
	tFile: PTFile,
	ctx: UploadCtx,
): Promise<string> {
	const key = await generateKey(binary, tFile, ctx.settings.s3.keyTemplate);
	await s3Upload(binary, key, tFile.extension, ctx);
	return ctx.settings.s3.publicUrl + key;
}

type TemplateParams =
	| "year"
	| "month"
	| "day"
	| "random2"
	| "random6"
	| "base62_of_ms_from_day_start"
	| "path"
	| "name"
	| "basename"
	| "extension";

export async function generateKey(
	binary: ArrayBuffer,
	tFile: PTFile,
	keyTemplate: string,
): Promise<string> {
	const params: Record<TemplateParams, string> = {
		year: new Date().getFullYear().toString(),
		month: (new Date().getMonth() + 1).toString().padStart(2, "0"),
		day: new Date().getDate().toString().padStart(2, "0"),
		random2: randomStringGenerator(2),
		random6: randomStringGenerator(6),
		base62_of_ms_from_day_start: encode62(
			Date.now() - new Date().setHours(0, 0, 0, 0),
		),
		path: tFile.path,
		name: tFile.name,
		basename: tFile.basename,
		extension: tFile.extension,
	};

	return template(keyTemplate, params);
}

function template(str: string, params: Record<string, string>): string {
	return str.replace(/\{\{(\w+)\}\}/g, (match, key: string) => {
		return params[key] || match;
	});
}

function randomStringGenerator(length: number) {
	const chars =
		"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
	let result = "";
	for (let i = 0; i < length; i++) {
		// use Math.random is OK here, as it's not for security purpose
		result += chars.charAt(Math.floor(Math.random() * chars.length));
	}
	return result;
}

async function s3Upload(
	file: ArrayBuffer | string,
	key: string,
	ext: string,
	ctx: UploadCtx,
) {
	const config = ctx.settings.s3;
	const client = new S3Client({
		region: config.region,
		forcePathStyle: config.forcePathStyle,
		credentials: {
			accessKeyId: config.accKeyId,
			secretAccessKey: config.secretAccKey,
		},
		endpoint: config.endpoint,
	});

	const command = new PutObjectCommand({
		Bucket: config.bucket,
		Key: key,
	});

	const url = await getSignedUrl(client, command);

	const resp = await ctx.requestUrl({
		url,
		method: "PUT",
		body: file,
		contentType: mime.getType(ext) || "application/octet-stream",
		throw: false,
	});

	if (resp.status !== 200) {
		throw new Error(`Failed to upload file: ${resp.status}`);
	}
}
