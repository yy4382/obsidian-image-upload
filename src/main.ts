import { Notice, Plugin, requestUrl, type TFile } from "obsidian";
import { DEFAULT_SETTINGS, SettingsTab, type Settings } from "./settings";
import { getDeletingFiles, transform } from "./transform";
import { upload } from "./upload";
import { createPromiseWithResolver, DeleteConfirmModal } from "./confirm-modal";

declare global {
	var cJS: () => Promise<unknown>;
}

export type PTFile = Pick<TFile, "basename" | "extension" | "name" | "path">;

export default class ImageUploadPlugin extends Plugin {
	settings: Settings;

	async onload() {
		await this.loadSettings();

		// This creates an icon in the left ribbon.
		this.addRibbonIcon("image-up", "Upload images for active file", () =>
			this.process(),
		);
		this.addCommand({
			id: "upload-image-for-active-file",
			name: "Upload images for active file",
			callback: () => this.process(),
		});

		// This adds a settings tab so the user can configure various aspects of the plugin
		this.addSettingTab(new SettingsTab(this.app, this));
	}

	onunload() {}

	async loadSettings() {
		this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
	}

	async saveSettings() {
		await this.saveData(this.settings);
	}

	async process() {
		const file = this.app.workspace.getActiveFile();
		if (!file) {
			new Notice("No active file");
			return;
		}
		const uploadedFiles = await this.uploadAndReplace(file);

		if (!uploadedFiles) {
			return;
		}

		new Notice(`Uploaded and replaced ${uploadedFiles.length} links`);

		if (!uploadedFiles.length) {
			return;
		}

		await this.deleteUploadedFiles(uploadedFiles, file);
	}

	async uploadAndReplace(file: TFile) {
		const cachedMetadata = this.app.metadataCache.getFileCache(file);
		if (!cachedMetadata) {
			new Notice("No assets in cache");
			return;
		}

		const { links, embeds } = cachedMetadata;

		if (!(links?.length || embeds?.length)) {
			new Notice("Nothing to upload");
			return;
		}

		const uploadedFiles = await asyncProcess(async (content: string) => {
			return transform(
				{
					settings: this.settings,
					uploader: await this.getUploader(),
					readBinary: (...args) => this.app.vault.readBinary(...args),
					resolveLink: (...args) =>
						this.app.metadataCache.getFirstLinkpathDest(...args),
					notice: (...args) => new Notice(...args),
				},
				content,
				{
					selfPath: file.path,
					links: links ?? [],
					embeds: embeds ?? [],
				},
			);
		}, file);

		return uploadedFiles;
	}
	async deleteUploadedFiles(uploadedFiles: string[], file: TFile) {
		const filesToDelete = getDeletingFiles(
			uploadedFiles,
			this.app.metadataCache.resolvedLinks,
			file.path,
		);

		console.log("Files to delete", filesToDelete);
		const { promise: confirmPromise, handler } = createPromiseWithResolver();
		new DeleteConfirmModal(this.app, handler, filesToDelete).open();
		try {
			await confirmPromise;
		} catch {
			new Notice("Deletion aborted");
			return;
		}

		for (const filePath of filesToDelete) {
			const file = this.app.vault.getFileByPath(filePath);
			if (!file) {
				console.warn("File not found", filePath);
				continue;
			}
			await this.app.vault.trash(file, this.settings.useSystemTrash);
		}
	}

	async getUploader(): Promise<
		(binary: ArrayBuffer, file: PTFile) => Promise<string>
	> {
		const { settings } = this;
		const className = settings.customUploaderClass;

		// If no class is provided, use the default S3 uploader
		if (!className) {
			return async (binary, file) => {
				return upload(binary, file, {
					settings,
					requestUrl: (...args) => requestUrl(...args),
				});
			};
		}
		let cJsObj: unknown;
		try {
			cJsObj = await cJS();
		} catch (e) {
			new Notice("Failed to load Custom JS");
			throw e;
		}

		if (!(cJsObj && typeof cJsObj === "object" && className in cJsObj)) {
			new Notice(`Class "${className}" not found in custom JS`);
			throw new Error(`Class "${className}" not found in custom JS`);
		}
		const uploaderClass = (cJsObj as Record<string, unknown>)[className];

		if (
			!(
				uploaderClass &&
				typeof uploaderClass === "object" &&
				"upload" in uploaderClass &&
				typeof uploaderClass.upload === "function"
			)
		) {
			new Notice(`Method "upload" not found in class "${className}"`);
			throw new Error(`Method "upload" not found in class "${className}`);
		}

		return async (binary, file) => {
			const result = await (
				uploaderClass as {
					upload: (binary: ArrayBuffer, file: PTFile) => Promise<unknown>;
				}
			).upload(binary, file);
			if (typeof result !== "string") {
				throw new Error("Result is not a string");
			}
			return result;
		};
	}
}

async function asyncProcess<T>(
	action: (content: string) => Promise<[string, T]>,
	file: TFile,
): Promise<T> {
	new Notice("Processing file. DO NOT MODIFY THE FILE UNTIL DONE");
	const content = await file.vault.cachedRead(file);
	const [newContent, other] = await action(content);
	await file.vault.process(file, (data) => {
		if (data !== content) {
			new Notice(
				"File has been updated during processing. User modified content will be concatenated to the end of the file",
			);
			return `${newContent}\n\`\`\`\n<!--Modified version while processing-->\n${data}\n\`\`\`\n`;
		}
		return newContent;
	});
	return other;
}
