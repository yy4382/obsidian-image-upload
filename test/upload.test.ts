import type { RequestUrlResponsePromise, requestUrl as ru } from "obsidian";
import { DEFAULT_SETTINGS } from "src/constants";
import { generateKey, upload, type UploadCtx } from "src/upload";
import { describe, test, expect, beforeEach, afterEach, vi } from "vitest";

const originalRandom = Math.random;

describe("generateKey", () => {
	beforeEach(() => {
		vi.useFakeTimers();
		const date = new Date(2024, 1, 1, 13);
		vi.setSystemTime(date);
		// mock Math.random
		const mockedRandom = vi.fn(() => 0.9);
		Math.random = mockedRandom;
	});

	afterEach(() => {
		vi.useRealTimers();

		Math.random = originalRandom;
	});
	test("should generate a key", async () => {
		const key = await generateKey(
			new ArrayBuffer(0),
			{
				basename: "basename",
				extension: "png",
				name: "basename.png",
				path: "path",
			},
			"{{year}}/{{month}}/{{day}}/{{random6}}/{{base62_of_ms_from_day_start}}-{{random2}}/{{name}}/{{basename}}.{{extension}}",
		);
		expect(key).toEqual("2024/02/01/333333/3amOI-33/basename.png/basename.png");
	});
	test("should generate a key", async () => {
		const key = await generateKey(
			new ArrayBuffer(0),
			{
				basename: "basename",
				extension: "png",
				name: "basename.png",
				path: "path",
			},
			"{{unknown}}",
		);
		expect(key).toEqual("{{unknown}}");
	});
});

describe("upload", () => {
	beforeEach(() => {
		vi.useFakeTimers();
		const date = new Date(2024, 1, 1, 13);
		vi.setSystemTime(date);
	});

	afterEach(() => {
		vi.useRealTimers();
	});
	test("upload", async () => {
		const binary = new Uint32Array([1, 2, 3]).buffer;
		const tFile = {
			basename: "basename",
			extension: "png",
			name: "basename.png",
			path: "basename.png",
		};

		let params = "";
		const requestUrl: typeof ru = (...args) => {
			params = JSON.stringify(args);
			return Promise.resolve({
				status: 200,
			}) as RequestUrlResponsePromise;
		};
		const ctx: UploadCtx = {
			settings: Object.assign(DEFAULT_SETTINGS, {
				s3: {
					bucket: "bucket",
					endpoint: "https://api.example.com",
					keyTemplate:
						"obsidian/{{year}}/{{month}}/{{day}}/{{base62_of_ms_from_day_start}}.{{extension}}",
					forcePathStyle: false,
					publicUrl: "https://example.com/",
					region: "auto",
					accKeyId: "Abc",
					secretAccKey: "Def",
				},
			}),
			requestUrl,
		};

		const url = await upload(binary, tFile, ctx);
		expect(url).toEqual("https://example.com/obsidian/2024/02/01/3amOI.png");
		expect(JSON.parse(params)).toMatchSnapshot();
	});
	test("upload with default mime", async () => {
		const binary = new Uint32Array([1, 2, 3]).buffer;
		const tFile = {
			basename: "basename",
			extension: "kkk",
			name: "basename.kkk",
			path: "basename.kkk",
		};

		let params = "";
		const requestUrl: typeof ru = (args) => {
			params = JSON.stringify(args);
			return Promise.resolve({
				status: 200,
			}) as RequestUrlResponsePromise;
		};
		const ctx: UploadCtx = {
			settings: Object.assign(DEFAULT_SETTINGS, {
				s3: {
					bucket: "bucket",
					endpoint: "https://api.example.com",
					keyTemplate:
						"obsidian/{{year}}/{{month}}/{{day}}/{{base62_of_ms_from_day_start}}.{{extension}}",
					forcePathStyle: false,
					publicUrl: "https://example.com/",
					region: "auto",
					accKeyId: "Abc",
					secretAccKey: "Def",
				},
			}),
			requestUrl,
		};

		const url = await upload(binary, tFile, ctx);
		expect(url).toEqual("https://example.com/obsidian/2024/02/01/3amOI.kkk");
		expect(JSON.parse(params)?.contentType).toEqual("application/octet-stream");
	});

	test("upload fail", async () => {
		const binary = new Uint32Array([1, 2, 3]).buffer;
		const tFile = {
			basename: "basename",
			extension: "png",
			name: "basename.png",
			path: "basename.png",
		};

		const requestUrl: typeof ru = (...args) => {
			return Promise.resolve({
				status: 400,
			}) as RequestUrlResponsePromise;
		};
		const ctx: UploadCtx = {
			settings: Object.assign(DEFAULT_SETTINGS, {
				s3: {
					bucket: "bucket",
					endpoint: "https://api.example.com",
					keyTemplate:
						"obsidian/{{year}}/{{month}}/{{day}}/{{base62_of_ms_from_day_start}}.{{extension}}",
					forcePathStyle: false,
					publicUrl: "https://example.com/",
					region: "auto",
					accKeyId: "Abc",
					secretAccKey: "Def",
				},
			}),
			requestUrl,
		};

		expect(async () => {
			await upload(binary, tFile, ctx);
		}).rejects.toThrowError("400");
	});

	test("upload with spaces in filename", async () => {
		const binary = new Uint32Array([1, 2, 3]).buffer;
		const tFile = {
			basename: "file with spaces",
			extension: "png",
			name: "file with spaces.png",
			path: "folder/file with spaces.png",
		};

		const requestUrl: typeof ru = (...args) => {
			return Promise.resolve({
				status: 200,
			}) as RequestUrlResponsePromise;
		};
		const ctx: UploadCtx = {
			settings: Object.assign(DEFAULT_SETTINGS, {
				s3: {
					bucket: "bucket",
					endpoint: "https://api.example.com",
					keyTemplate: "{{name}}",
					forcePathStyle: false,
					publicUrl: "https://example.com/",
					region: "auto",
					accKeyId: "Abc",
					secretAccKey: "Def",
				},
			}),
			requestUrl,
		};

		const url = await upload(binary, tFile, ctx);
		expect(url).toEqual("https://example.com/file%20with%20spaces.png");
	});
});
