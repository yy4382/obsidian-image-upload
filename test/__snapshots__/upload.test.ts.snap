// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`upload > upload 1`] = `
[
  {
    "body": {},
    "contentType": "image/png",
    "method": "PUT",
    "throw": false,
    "url": "https://bucket.api.example.com/obsidian/2024/02/01/3amOI.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=Abc%2F20240201%2Fauto%2Fs3%2Faws4_request&X-Amz-Date=20240201T210000Z&X-Amz-Expires=900&X-Amz-Signature=855d552294c58a9eb91a2d9ed21abc224a8a8e35568e6e7423e71b3b8e34e2fc&X-Amz-SignedHeaders=host&x-id=PutObject",
  },
]
`;
